pcapgo/capture.go

```diff
@@ -0,0 +1,48 @@
+package pcapgo
+
+import (
+	"net"
+	"syscall"
+	"time"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"github.com/mdlayher/raw"
+)
+
+// EthernetHandle wraps a raw.Conn, implementing gopacket.PacketDataSource so
+// that the handle can be used with gopacket.NewPacketSource.
+type EthernetHandle struct {
+	*raw.Conn
+}
+
+// ReadPacketData implements gopacket.PacketDataSource.
+func (h *EthernetHandle) ReadPacketData() ([]byte, gopacket.CaptureInfo, error) {
+	b := make([]byte, 4096) // TODO(correctness): how much space do we need?
+	n, _, err := h.ReadFrom(b)
+	if err != nil {
+		return nil, gopacket.CaptureInfo{}, err
+	}
+	packet := gopacket.NewPacket(b[:n], layers.LayerTypeEthernet, gopacket.Default)
```

Comment:
Don't create a new packet and return it.  Just return b[:n], and gopacket's https://github.com/google/gopacket/blob/master/packet.go#L794 will wrap it into a packet.
With the current code, a full packet decode happens, then the entire packet object is thrown out.

Reply:
Whoops, of course. Done.